tox.ini

```diff
@@ -32,16 +32,16 @@ commands =
 
 [testenv:fmt]
 description = Apply coding style standards to code
-deps =
-    -r{toxinidir}/requirements-dev.txt
+deps = pip-tools
+commands_pre = pip-sync {toxinidir}/requirements-dev.txt
```

Comment:
I'm happy to include this change - I agree that it's probably better. It does mean there's a bit of duplication, ie. I've got this now:

```yaml
[project.optional-dependencies]
docs = [
    "sphinx==6.2.1",
    "sphinx-design",
    "furo",
    "sphinx-tabs",
    "lxd-sphinx-extensions",
    "sphinx-copybutton",
]
lint = [
    "autopep8~=1.6",
    "flake8~=6.1",
    "flake8-docstrings~=1.7",
    "flake8-builtins~=2.1",
    "isort~=5.11",
    "pep8-naming~=0.13",
    "pyproject-flake8~=6.1",
]
static = [
    "pyright==1.1.345",
    "pytest~=7.2",
    "typing_extensions~=4.2",
]
unit = [
    "coverage[toml]~=7.0",
    "pytest~=7.2",
    "typing_extensions~=4.2",
]
smoke = [
    "coverage[toml]~=7.0",
    "pytest~=7.2",
    "pytest-operator~=0.23",
]
```

(I have tested all except smoke).

I could group pytest, typing_extensions, and coverage into "test" but I don't think there's a way to say "this group requires this other group as well" - obviously we can handle that in `tox.ini` for our use and anyone just running from tox, but maybe it becomes unclear to other people?

Or I could decrease the number of groups (I already have fmt and lint use the same one), so static, unit, and smoke all use the same one? `static` needs `pytest` because we type check the tests, I think.

Reply:
Let's keep them separate and simple for now (even if a little bit of duplication). We can always tweak it later.

Out of interest, why does `static` require pytest and typing_extensions?

Reply:
> Out of interest, why does `static` require pytest and typing_extensions?

`typing_extensions` because `static` is checking types :smile: - it's the library that backports Python typing to older Python versions (we only use `Required` and `NotRequired` in `TypedDict`s currently, I think). I believe `unit` needs it as well because pytest (or unittest) sets `typing.TYPE_CHECKING` to True.

`static` needs `pytest` not for the tool but because it's imported (in the tests, which we are also type-checking). These build on top of the main dependencies, so the core ones are already there.